lib/ecto/uuid.ex

```diff
@@ -13,8 +13,14 @@ defmodule Ecto.UUID do
   @doc """
   Casts to UUID.
   """
-  def cast(<< _::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96 >> = u), do: {:ok, u}
-  def cast(_), do: :error
+  def cast(uuid) do
+    regex = ~r/\A[a-f0-9]{8}-[a-f0-9]{4}-[1-5][a-f0-9]{3}-[89ab][a-f0-9]{3}-[a-f0-9]{12}\z/i
```

Comment:
Please don't use a regular expression. :) See the function below on this module on how you can use pattern matching to solve this. I assume it will end-up looking like dump but instead of converting to the underlying binary, we will keep the normalized UUID.